### DISCUSSION TITLE
How to get time from server

### DISCUSSION THREAD

**Discussion by Arcanum7** [Lora-net/LoRaMac-node#1171] [Q&A]
Hi! I have sx1272 + stm32L072 and use loramac lib.
How can i get UTC time fom server on every TXRX frame by MAC commands?

**Comment by mluis1** [comment]
The LoRaWAN specification defines a MAC command named `DeviceTimeReq/DeviceTimeAns`.

Therefore in order to get the time from the Network Server the end-device needs to ask the execution of a `DeviceTimeReq` MAC command.

An example on how to do this can be found below
https://github.com/Lora-net/LoRaMac-node/blob/05353ff70bec5349a26b23a02168accb0ea03ad9/src/apps/LoRaMac/common/LmHandler/LmHandler.c#L520-L539

Please note that the received time will be based on GPS epoch. 
This code also provides the time as UTC based (without leap seconds).
You need to implement a way to convey the current number of leap seconds. If this is not implemented the time will be wrong by 27 seconds (as of today).

Your second sentence is confusing. You should not query for time at every transmission. You should do this only once in a while.
